stdlib/public/core/IntegerParsing.swift

```diff
@@ -146,31 +43,37 @@ extension FixedWidthInteger {
   ///     `radix`.
   ///   - radix: The radix, or base, to use for converting `text` to an integer
   ///     value. `radix` must be in the range `2...36`. The default is 10.
-  @inlinable // @specializable
-  @_semantics("optimize.sil.specialize.generic.partial.never")
+  @inlinable @inline(__always)
   public init?<S: StringProtocol>(_ text: S, radix: Int = 10) {
-    _precondition(2...36 ~= radix, "Radix not in range 2...36")
-
-    if let str = text as? String, str._guts.isFastUTF8 {
-      guard let ret = str._guts.withFastUTF8 ({ utf8 -> Self? in
-        var iter = utf8.makeIterator()
-        return _parseASCII(codeUnits: &iter, radix: Self(radix))
-      }) else {
-        return nil
-      }
-      self = ret
-      return
+    precondition(2...36 ~= radix)
+    let wholeGuts = text._wholeGuts
+    // The specialized paths require that all of the contiguous bytes can
+    // be read using UInt64 loads from (address & ~7).
+    if wholeGuts._object.isPreferredRepresentation, Self.bitWidth <= 64 {
```

Comment:
@milseman Would this be the correct check to guarantee `_loadUnalignedChunk` is valid? It's clear valid for small strings, and AFAICT also for natively stored large strings (both on 32 & 64-bits), not so sure about literals.

Reply:
Or, if we can perhaps even guarantee this for shared and/or bridged UTF-8?

Reply:
I don't think we can make this assumption. At the very least, a `CFString` may come in unaligned. CC @stephentyrone . We want better utilities for dealing with larger-strides alignment in general.

Reply:
Note that the requirement isn't alignment exactly, but rather that the memory at the address rounded down to a 8-byte multiple (even on 32-bit systems) is valid to read even if it's garbage.

Per string backing:
- Small: should be fine
- Large (native, mortal): should be fine on both 32 & 64 bit AFAICT
- Large, immortal: not sure about this
- Shared & bridged: Probably not?
- Foreign: no

Reply:
This assumption does not hold. Swift hasn't yet defined a formal memory model that addresses this detail, but absent a guarantee, it is UB to read outside the bounds of an object, *even if that's an aligned read, even if it would be allowed by the hardware*. If we want to take advantage of hardware support for doing this, it has to be done in assembly at present (C and C++ make this explicitly UB, Swift may or may not in the future, so we can't depend on the semantics at this point).

Reply:
For objects in general yes, but for some specific string backing it should be valid. In particular (correct me if I'm wrong):
- Small: UTF-8 pointer points to 2 UInt64's on the stack (on both 32 & 64-bit arch), so the assumption holds
- Large (native, mortal): UTF-8 pointer is tail allocated memory with header of 20/32 bytes (32/64 bits) and for <128 bytes the capacity is bucketed to 16 byte multiples. Above that it should have breadcrumbs, which is word size. That only leaves a question for 32-bit platforms?
Not so sure about the memory layout of literals.

Reply:
The most robust approach here would be to take the fast(est) path if the pointer is suitably aligned, else fall back to a byte-by-byte approach.  In particular, that will be safe even if the string internals change in the future.

Reply:
If it's suitably aligned, then it's valid to read from, but then there's no need to round down, because it's suitably aligned.

Reply:
@stephentyrone does the above reasoning for small & native string not hold then? I'm trying to find a subset for which we can have this guarantee because the performance difference is rather large, and could be generalized to other string operations such as upper/lowercasing.

@tbkka Note there's also the question of reading beyond the end. And the perf difference between byte-by-byte approach and chunk-by-chunk is quite large so if at all possible it'd be worth it.

Reply:
@PatrickPijnappel There's no need to go byte-by-byte or read beyond the end; both misalignment and trailing bytes can be handled without significant performance penalty and without reading any memory outside the string bounds:

- If small, just load two words and check them, masking out unused bytes with a shift.
- If 8 bytes or less, and not small, go byte by byte (this case is quite rare, perf doesn't matter much)
- Otherwise:
  - Load the first 8 bytes of the string, regardless of alignment. This is not quite expressible in Swift today, so we'll have to define an unaligned load function in the stubs header that does it (this is trivial, happily).
  - Advance to the next aligned location in the string, `start+8 & -8`. Subtract the distance from this pointer and `start` from length.
  - As long as length > 8, load aligned words.
  - Do a final unaligned load from `end-8`.

This does two unaligned loads, but all other accesses are aligned, and no byte-by-byte access, and never reads outside the valid string bounds. On modern x86 and arm implementations the two unaligned accesses are fast (and the first one will actually be aligned in practice almost every time).

Reply:
That makes a lot of sense, I'll implement that this weekend! Can you give some pointers to adding the unaligned load to the stub header?

Reply:
@PatrickPijnappel -- I think @stephentyrone is referring to this header file: `stdlib/public/SwiftShims/RuntimeShims.h` and the associated implementations which seem to be mostly in `stdlib/public/stubs/Stubs.cpp`.  These files (and the other files in those directories) provide various small C/C++ support functions to the standard library, many of which serve to hide processor- or operating-system-specific details.

Reply:
Do you know whether this should be an inline function or whether that doesn't matter for the compiler? The stubs seem to be defined a couple of different places, e.g. RuntimeStubs.h, RuntimeShims.h and Runtime.swift (in stdlib), which is the proper place?

Implementation-wise do we just need
```cpp
uint64_t loadUInt64Unaligned(void *p) {
  return *(uint64_t *)p;
}
```

Reply:
No, that generates an aligned load; by casting the pointer to `uint64_t`, you are telling the compiler that it is suitably aligned for an 8-byte integer.

You want:
```c
static inline SWIFT_ALWAYS_INLINE
__swift_uint64_t loadUInt64Unaligned(char *p) {
  __swift_uint64_t result;
  memcpy(&result, p, sizeof result);
  return result;
}
```

Reply:
@stephentyrone The `memcpy` appears to hit some bug in SIL deserialization: https://gist.github.com/PatrickPijnappel/f133f7296eb05b966a7bc9486ded35b7

Implemented it using the alignment attribute for now

Reply:
Ah, right, forgot that you can't use `stdlib` functions in the shims. If you use `__builtin_memcpy` instead it will work.